Add basic support for MP4 files

vrischmann opened the pull request:
This PR adds basic support for reading MP4 files.

I only cared about reading metadata so the code is extremely simple and makes some assumptions about the layout of "atoms" in files.
I've added a test mp4 file with all tags that EasyTag allowed me to set.

I don't have a lot of "real-world" mp4 files, only some that I got with youtube-dl, so it's possible something is broken.

squeek502 commented on src/mp4.zig:
```diff
@@ -0,0 +1,212 @@
+// zig fmt: on
+
+fn getMetadataAtom(name: []const u8) ?MetadataAtom {
+    for (metadata_atoms) |atom| {
```
This can be an `inline for` I believe (alternatively, `metadata_atoms` could use `std.ComptimeStringMap` but given that all the names are the same length, it won't have much benefit over `inline for`).

vrischmann replied:
Yep, done

squeek502 reviewed (commented):
Thanks! Code looks good on first glance. I'd like to familiarize myself a bit more with the mp4 format and do some cursory fuzz testing before merging.

vrischmann commented:
> I'd like to familiarize myself a bit more with the mp4 format and do some cursory fuzz testing before merging.

  Sure ! (also thanks for your articles on fuzz testing, it helped me a lot when working on my sqlite wrapper)

Note that I'm not that familiar with MP4 either, I learned what I know over the week-end so it's entirely possible I missed something.

squeek502 commented on src/mp4.zig:
```diff
@@ -0,0 +1,212 @@
+            // a size of 1 means the atom header has an extended size field
+            // TODO: implement this if relevant ?
+
+            std.debug.panic("unhandled 64 bit size", .{});
```
The goal is to never induce a panic regardless of the input, so this should be something like:

```suggestion
            // TODO: implement this if relevant ?

            return error.UnimplementedSizeFormat;
```

squeek502 commented on src/mp4.zig:
```diff
@@ -0,0 +1,212 @@
+
+/// Generic data atom
+///
+/// Its value should be interpreted based on the type indicator and locale indicator.
+/// For our purposes it's probably enough to treat them as UTF-8 strings.
```
I think it might make more sense to validate that the data's type is UTF-8 and skip anything that's not explicitly UTF-8. AFAICT that would mean verifying that the first byte of the type indicator is `0` and that the remaining `u24` is `1`.

- [Type Indicator](https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/Metadata/Metadata.html#//apple_ref/doc/uid/TP40000939-CH1-SW28)
- [Well-Known Types](https://developer.apple.com/library/archive/documentation/QuickTime/QTFF/Metadata/Metadata.html#//apple_ref/doc/uid/TP40000939-CH1-SW34)

EDIT: Accepting Well-Known Type `0` might also make sense (assuming it's used for ASCII data? Not sure what the common case is)

vrischmann replied:
I implemented this.

As far as I can tell with my mp4 files the well-known type is always 1 for metadata.

squeek502 commented on src/mp4.zig:
```diff
@@ -0,0 +1,212 @@
+
+        // Skip every atom we don't recognize or are not interested in.
+
+        try seekable_stream.seekBy(atom_header.size - 8);
```
This subtraction can overflow if the header size is set as less than 8 (an invalid header size). Adding

```zig
    if (size < 8) {
        return error.AtomSizeTooSmall;
    }
```
after the `const size = switch (try reader.readIntBig(u32)) { ... };` in  `readAtomHeader` would be one possible fix.

squeek502 reviewed (changes requested):
Did some fuzzing and ran into a few issues (mentioned in review comments). Here's some test code to induce them (can be added to the bottom of `mp4.zig`):

```zig
fn testFuzzedData(allocator: Allocator, data: []const u8) !void {
    var stream_source = std.io.StreamSource{ .const_buffer = std.io.fixedBufferStream(data) };
    // any non-panic is acceptable
    _ = read(allocator, stream_source.reader(), stream_source.seekableStream()) catch {};
}

test "fuzzed inputs" {
    const allocator = std.testing.allocator;

    // integer overflow when skipping an unrecognized atom
    try testFuzzedData(allocator, "\x00\x00\x00\x00\xe6\x95\xbe");

    // unhandled 64 bit size
    try testFuzzedData(allocator, "ID3\x04\x00\x04\x00\x00\x00\x00\x00\x00\x00\x01\xfe\xff\x00N\x00o\x00 \x00I\x00s\x00l\x00a\x00n\x00d\x00 1TX8t Test[X\x00\x00\x00\x001TXXt TestTXXX\x00\x00\x00\x13\x00\x00\x03Test8 \xf2\x95\xbb\xe5\x13\x00\x00\x03T\xe6\x94\xbe");
}
```

There's also the possibility of trying to read impossibly-large data (larger than the filesize) in `readDataAtom`. Here's a zipped up file that can reproduce it:

[impossibly-large-read.zip](https://github.com/squeek502/audiometa/files/8517206/impossibly-large-read.zip)

It'll try to allocate / read a value of size `808464416` which should be rejected as it is larger than the remaining data. See [here](https://github.com/squeek502/audiometa/blob/5de0df0d3e5c1c493ab1712bccdfe73b38a0bea2/src/id3v2.zig#L501-L504) for where this type of thing is guarded against in the ID3v2 code.

vrischmann commented:
```zig
 // unhandled 64 bit size
    try testFuzzedData(allocator, "ID3\x04\x00\x04\x00\x00\x00\x00\x00\x00\x00\x01\xfe\xff\x00N\x00o\x00 \x00I\x00s\x00l\x00a\x00n\x00d\x00 1TX8t Test[X\x00\x00\x00\x001TXXt TestTXXX\x00\x00\x00\x13\x00\x00\x03Test8 \xf2\x95\xbb\xe5\x13\x00\x00\x03T\xe6\x94\xbe");
```

this doesn't actually trigger the panic for me, but I added a proper error and test.
I addressed the header size check.

I'll check the other 2 issues.

vrischmann commented:
I think I addressed all your feedback. I also fixed an additional overflow bug found via fuzzing.

squeek502 commented:
Looks good, thanks! I only just realized that `.m4a` is just another extension for the same format, and I have a non-zero amount of `.m4a` files in my music collection, so I'll likely be expanding on this implementation in the future.

(also, thanks for your `zig-sqlite` library, I used it recently when making [this](https://github.com/squeek502/watchedoverlay) and it made that part of things very easy)
